aea/helpers/dialogue/base.py

```diff
@@ -412,185 +403,361 @@ def last_message(self) -> Optional[Message]:
+
+        :param message_id: the id of the message
         :return: the message if it exists, None otherwise
         """
         result = None  # type: Optional[Message]
         list_of_all_messages = self._outgoing_messages + self._incoming_messages
         for message in list_of_all_messages:
-            if message.message_id == message_id_to_find:
+            if message.message_id == message_id:
                 result = message
                 break
-
         return result
 
-    @property
-    def is_empty(self) -> bool:
+    def get_message(self, message_id: int) -> Message:
         """
-        Check whether the dialogue is empty.
+        Get the message whose id is 'message_id'.
 
-        :return: True if empty, False otherwise
+        :param message_id: the id of the message
+        :return: the message
+        :raises: AssertionError if message is not present
         """
-        return len(self._outgoing_messages) == 0 and len(self._incoming_messages) == 0
+        message = self.try_get_message(message_id)
+        assert message is not None, "Message not present."
+        return message
 
-    def update(self, message: Message) -> bool:
+    def has_message(self, message: Message) -> bool:
         """
-        Extend the list of incoming/outgoing messages with 'message', if 'message' belongs to dialogue and is valid.
+        Check whether a message exists in this dialogue.
 
-        :param message: a message to be added
-        :return: True if message successfully added, false otherwise
+        :param message: the message
+        :return: True if message exists in this dialogue, False otherwise
         """
-        if not message.is_incoming:
-            message.sender = self.agent_address
+        if self.is_empty:
+            return False
 
-        self.ensure_counterparty(message)
+        if not self.has_message_id(message.message_id):
+            return False
 
-        if not self.is_belonging_to_dialogue(message):
+        retrieved_message = self.get_message(message.message_id)
+        return message == retrieved_message
+
+    def has_message_id(self, message_id: int) -> bool:
+        """
+        Check whether a message with the supplied message id exists in this dialogue.
+
+        :param message_id: the message id
+        :return: True if message with that id exists in this dialogue, False otherwise
+        """
+        if self.is_empty:
             return False
 
-        is_extendable = self.is_valid_next_message(message)
-        if is_extendable:
-            if message.is_incoming:
-                self._incoming_messages.extend([message])
-            else:
-                self._outgoing_messages.extend([message])
-        return is_extendable
+        return self.STARTING_MESSAGE_ID <= message_id <= self.last_message.message_id  # type: ignore
 
-    def ensure_counterparty(self, message: Message) -> None:
+    def update(self, message: Message) -> None:
```

Comment:
We should make this method private (i.e. relabel to `_update`). In the new API concept, the `update` functionality is either called from `reply` or from the `Dialogues`' `create` and `reply`, we should avoid it being called directly. By making this private, we can also maintain the old method `update` with its old signature (returning a bool) (which wraps `_update` in a try, except block). This makes upgrading easier and we can furnish the old `update` with a `DeprecationWarning`. This change should be done after we have converted all skills to new API (so not as part of this PR).

Reply:
Old update can then be removed fully with `0.7.0`

Reply:
I agree with privatising update (`update` to `_update`) to discourage use.

I don't quite like postponing the cleanups to another version and introduce interim functions. Developers cannot avoid changing their code if they want to upgrade to 0.6.0 (which is the only distinction between major/minor versions which we placed ourselves, i.e. backwards compatibility). 

I don't see why them not upgrading one part of their code in 0.6.0 (`Dialogue.update`) and doing it in 0.7.0 is helpful. On the other side, we just have to come back to this part of the code again in a few weeks and make the move completely.